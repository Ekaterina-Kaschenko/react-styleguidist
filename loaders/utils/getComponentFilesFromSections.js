'use strict';

const getComponentFiles = require('./getComponentFiles');

/**
 * Return absolute paths of all components in sections.
 *
 * @param {Array} sections
 * @param {string} rootDir
 * @returns {Array}
 */
module.exports = function getComponentFilesFromSections(sections, rootDir) {
	return sections.reduce((components, section) => {
		if (section.components) {
			return components.concat(getComponentFiles(section.components, rootDir));
		}

		if (section.sections) {
			return components.concat(getComponentFilesFromSections(section.sections, rootDir));
		}

		return components;
	}, []);
};
