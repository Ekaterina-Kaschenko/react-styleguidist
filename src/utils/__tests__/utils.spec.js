const map = require('lodash/map');
import * as utils from '../utils';

const COMPONENTS = [
	{
		name: 'Button',
	},
	{
		name: 'Image',
	},
	{
		name: 'Input',
	},
	{
		name: 'Link',
	},
	{
		name: 'Textarea',
	},
];

const SECTIONS = [
	{
		name: 'General',
		sections: [
			{
				name: 'Particles',
				components: [
					{
						name: 'Button',
					},
					{
						name: 'Image',
					},
				],
			},
		],
	},
];

let sourceGlobalLength;
beforeEach(() => {
	sourceGlobalLength = Object.keys(global).length;
});
afterEach(() => {
	delete global.Foo;
	delete global.Bar;
});

describe('setComponentsNames', () => {
	it('should set name property to each component', () => {
		const result = utils.setComponentsNames([
			{
				module: {
					displayName: 'Foo',
				},
			},
			{
				module: {
					name: 'Bar',
				},
			},
			{
				module: {
					displayName: 'Foo',
				},
				props: {
					displayName: 'FooOverride',
				},
			},
		]);
		expect(map(result, 'name')).toEqual(['Foo', 'Bar', 'FooOverride']);
	});
});

describe('setSlugs', () => {
	it('should set unique slug property to each section or component with name property', () => {
		const result = utils.setSlugs([
			{
				name: 'Multiple words',
				sections: [
					{
						name: 'Single',
						components: [
							{
								name: 'Button',
							},
							{
								name: 'Image',
							},
						],
					},
					{
						components: [
							{
								name: 'Button',
							},
							{
								name: 'Image',
							},
						],
					},
				],
			},
		]);
		expect(result).toMatchSnapshot();
	});
});

describe('globalizeComponents', () => {
	it('should set each componentâ€™s module as a global variable', () => {
		utils.globalizeComponents([
			{
				name: 'Foo',
				module: 13,
			},
			{
				name: 'Bar',
				module: 27,
			},
			{
				name: 'PathedFoo',
				module: { a: 32 },
				props: {
					path: 'a',
				},
			},
		]);
		expect(Object.keys(global).length).toBe(sourceGlobalLength + 3);
		expect(global.Foo).toBe(13);
		expect(global.Bar).toBe(27);
		expect(global.PathedFoo).toBe(32);
	});
});

describe('getFilterRegExp', () => {
	it('should return a RegExp', () => {
		const result = utils.getFilterRegExp('');
		expect(result instanceof RegExp).toBe(true);
	});

	it('RegExp should fuzzy match a string', () => {
		const result = utils.getFilterRegExp('btn');
		expect('button').toMatch(result);
	});

	it('RegExp should not match when string is different', () => {
		const result = utils.getFilterRegExp('buttons');
		expect('button').not.toMatch(result);
	});

	it('should not throw when query contains special characters', () => {
		const fn = () => utils.getFilterRegExp('\\');
		expect(fn).not.toThrow();
	});

	it('RegExp should ignore non-alphanumeric characters', () => {
		const result = utils.getFilterRegExp('#$b()tn');
		expect('button').toMatch(result);
	});
});

describe('filterComponentsByName', () => {
	it('should return initial list with empty query', () => {
		const result = utils.filterComponentsByName(COMPONENTS, '');
		expect(result).toEqual(COMPONENTS);
	});

	it('should return filtered list, should ignore case', () => {
		const result = utils.filterComponentsByName(COMPONENTS, 'button');
		expect(result).toEqual([{ name: 'Button' }]);
	});

	it('should return empty list when nothing found', () => {
		const result = utils.filterComponentsByName(COMPONENTS, 'pizza');
		expect(result).toEqual([]);
	});
});

describe('filterComponentsByExactName', () => {
	it('should return components with exact name', () => {
		const result = utils.filterComponentsByExactName(COMPONENTS, 'Image');
		expect(result).toEqual([COMPONENTS[1]]);
	});
});

describe('filterComponentsInSectionsByExactName', () => {
	it('should return components at any level with exact name', () => {
		const result = utils.filterComponentsInSectionsByExactName(SECTIONS, 'Image');
		expect(result).toEqual([COMPONENTS[1]]);
	});
});

describe('getComponentNameFromHash', () => {
	it('should return important part of hash if it contains component name', () => {
		const result = utils.getComponentNameFromHash('#!/Button');
		expect(result).toEqual({ targetComponentName: 'Button', targetComponentIndex: null });
	});

	it('should return an empty object if hash contains no component name', () => {
		const result = utils.getComponentNameFromHash('Button');
		expect(result).toEqual({});
	});
});

describe('filterComponentExamples', () => {
	it('should return a shallow copy of the component with example filtered by given index', () => {
		const comp = { examples: ['a', 'b', 'c', 'd'], other: 'info' };
		const expectedOutput = { examples: ['c'], other: 'info' };
		const result = utils.filterComponentExamples(comp, 2);
		expect(result).toEqual(expectedOutput);
	});
});
