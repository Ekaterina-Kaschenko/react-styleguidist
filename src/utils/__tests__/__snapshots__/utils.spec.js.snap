exports[`setSlugs should set unique slug property to each section or component with name property 1`] = `
Array [
  Object {
    "name": "Multiple words",
    "sections": Array [
      Object {
        "components": Array [
          Object {
            "name": "Button",
            "slug": "button",
          },
          Object {
            "name": "Image",
            "slug": "image",
          },
        ],
        "name": "Single",
        "slug": "single",
      },
      Object {
        "components": Array [
          Object {
            "name": "Button",
            "slug": "button-1",
          },
          Object {
            "name": "Image",
            "slug": "image-1",
          },
        ],
      },
    ],
    "slug": "multiple-words",
  },
]
`;
